File: python/pytoon/normalize.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from datetime import date, datetime
from decimal import Decimal
from typing import Any

from pytoon.types import JsonArray, JsonObject, JsonPrimitive, JsonValue


def normalize_value(value: Any) -> JsonValue:
    if value is None:
        return None
    if isinstance(value, (str, bool)):
        return value
    if isinstance(value, (int, float)):
        if isinstance(value, float):
            if value != value or value in {float('inf'), float('-inf')}:
                return None
            if value == 0.0:
                return 0
        return int(value) if isinstance(value, bool) else value
    if isinstance(value, Decimal):
        if not value.is_finite():
            return None
        if value == value.to_integral_value() and value.as_tuple().exponent >= 0:
            as_int = int(value)
            if -(2**53) < as_int < 2**53:
                return as_int
        return float(value)
    if isinstance(value, (datetime, date)):
        return value.isoformat()
    if isinstance(value, set):
        return [normalize_value(item) for item in value]
    if isinstance(value, Sequence) and not isinstance(value, (str, bytes, bytearray)):
        return [normalize_value(item) for item in value]
    if isinstance(value, Mapping):
        return {str(key): normalize_value(val) for key, val in value.items()}
    if hasattr(value, '__dict__'):
        return {str(key): normalize_value(val) for key, val in vars(value).items()}
    return None


def is_json_primitive(value: Any) -> bool:
    return value is None or isinstance(value, (str, int, float, bool))


def is_json_array(value: Any) -> bool:
    return isinstance(value, list)


def is_json_object(value: Any) -> bool:
    return isinstance(value, dict)


def is_empty_object(value: JsonObject) -> bool:
    return not value


def is_array_of_primitives(value: JsonArray) -> bool:
    return all(is_json_primitive(item) for item in value)


def is_array_of_arrays(value: JsonArray) -> bool:
    return all(isinstance(item, list) for item in value)


def is_array_of_objects(value: JsonArray) -> bool:
    return all(isinstance(item, dict) for item in value)

File: packages/toon_julia/src/encode/normalize.jl
"""
Normalizes a value to JsonValue type.
Handles various Julia types and converts them to JSON-compatible types.
"""
function normalize_value(value)::JsonValue
    # null/nothing
    if value === nothing
        return nothing
    end
    
    # Primitives
    if value isa String || value isa Bool
        return value
    end
    
    # Numbers
    if value isa Number
        if !isfinite(value)
            return nothing
        end
        # Normalize -0 to 0
        if value isa Float64 && value == 0.0 && signbit(value)
            return 0.0
        end
        return value
    end
    
    # Array
    if value isa AbstractArray
        return Any[normalize_value(item) for item in value]
    end
    
    # Dict/AbstractDict
    if value isa AbstractDict && !(value isa AbstractArray)
        normalized = Dict{String, Any}()
        for (k, v) in value
            normalized[string(k)] = normalize_value(v)
        end
        return normalized
    end
    
    # Fallback: function, symbol, or other → nothing
    return nothing
end

# Type guards
function is_json_primitive(value)::Bool
    return value === nothing || value isa String || value isa Number || value isa Bool
end

function is_json_array(value)::Bool
    return value isa AbstractArray
end

function is_json_object(value)::Bool
    return value isa AbstractDict && !(value isa AbstractArray) && !(value isa Tuple)
end

function is_empty_object(value::JsonObject)::Bool
    return isempty(value)
end

function is_plain_dict(value)::Bool
    return value isa Dict
end

# Array type detection
function is_array_of_primitives(value::JsonArray)::Bool
    return isempty(value) || all(item -> is_json_primitive(item), value)
end

function is_array_of_arrays(value::JsonArray)::Bool
    return isempty(value) || all(item -> is_json_array(item), value)
end

function is_array_of_objects(value::JsonArray)::Bool
    return isempty(value) || all(item -> is_json_object(item), value)
end

